syne_tune/optimizer/schedulers/transfer_learning/rush.py

```diff
@@ -0,0 +1,159 @@
+        in addition to top performing configurations from other tasks and also serve to preemptively prune
+        underperforming configurations
+        """
+        super().__init__(config_space=config_space,
+                         transfer_learning_evaluations=transfer_learning_evaluations,
+                         metric_names=[metric])
+        threshold_candidates = RUSHScheduler._compute_best_hyperparameter_per_task(
+            config_space=config_space,
+            transfer_learning_evaluations=transfer_learning_evaluations,
+            metric=metric,
+            mode=hyperband_kwargs.get('mode', 'min')
+        )
+        if points_to_evaluate is not None:
+            threshold_candidates += points_to_evaluate
+            threshold_candidates = [dict(s) for s in set(frozenset(p.items()) for p in threshold_candidates)]
+
+        self._hyperband_scheduler = HyperbandScheduler(config_space, metric=metric,
+                                                       points_to_evaluate=threshold_candidates, **hyperband_kwargs)
+        assert self._hyperband_scheduler.scheduler_type in ['promotion', 'stopping'],\
+            "RUSH supports only type 'stopping' or 'promotion'"
+        self._num_init_configs = len(threshold_candidates)
+        self._thresholds = dict()  # thresholds at different resource levels that must be met
+
+    @staticmethod
+    def _compute_best_hyperparameter_per_task(config_space: Dict,
+                                              transfer_learning_evaluations: Dict[str, TransferLearningTaskEvaluations],
+                                              metric: str,
+                                              mode: str) -> List[Dict]:
+        argbest, best = (np.argmin, np.min) if mode == 'min' else (np.argmax, np.max)
+        baseline_configurations = list()
+        for evals in transfer_learning_evaluations.values():
+            best_hpc_idx = argbest(best(evals.objective_values(objective_name=metric)[..., -1], axis=1))
+            hpc = evals.hyperparameters.iloc[best_hpc_idx]
+            baseline_configurations.append({
+                key: hpc[key] for key in config_space
+            })
+        return baseline_configurations
+
+    def on_trial_error(self, trial: Trial) -> None:
+        self._hyperband_scheduler.on_trial_error(trial)
+
+    def on_trial_result(self, trial: Trial, result: Dict) -> str:
+        trial_decision = self._hyperband_scheduler.on_trial_result(trial, result)
+        return self._on_trial_result(trial_decision, trial, result)
+
+    def on_trial_remove(self, trial: Trial) -> None:
+        self._hyperband_scheduler.on_trial_remove(trial)
+
+    def on_trial_complete(self, trial: Trial, result: Dict) -> None:
+        self._hyperband_scheduler.on_trial_complete(trial, result)
+
+    def suggest(self, trial_id: int) -> Optional[TrialSuggestion]:
+        return self._hyperband_scheduler.suggest(trial_id)
```

Comment:
The fact that you simply call suggest here, shows that something is missing. In type='promotion', the decision is made in suggest (whether to promote a paused trial or not).

It would be really hard to hack this post-hoc. Say normal ASHA decides to promote a trial here, but then RUSH says: no. Then, you'd have to go back to ASHA and ask "would you want to promote another trial, but not that one here?". This is not supported.

And if RUSH would block all possible ASHA promotions, then you'd have to call the searcher to suggest a new config.

This is too complex. Better implement the RungSystem.